### DISCUSSION TITLE
Getting address of image from imgfetch

### DISCUSSION THREAD

**Discussion by tyne2022** [ipxe/ipxe#588] [Q&A]
I am using iPXE with an embedded system that uses U-Boot. I am able to build iPXE as a EFI image for ARM64 and load it from U-Boot.

I would like to download an FPGA bitstream (generic binary file) over HTTPS with iPXIE, get the address the image is held at, exit iPXE and return to U-Boot, and finally configure the FPGA with the bitstream in memory.

How can I get the address of an image (and size) of an image from imgfetch? Is there a mechanism to pass this back to U-Boot?

For example, in iPXE, I could do:
```
dhcp
imgfetch https://server/fpga.bin
exit
```

If I knew that `fpga.bin` was loaded to `0xdeadbeef` with size `0xcafe` by iPXE, in U-Boot, I could then do:
```
fpga load 0 0xdeadbeef 0xcafe
```

Thanks for the help!

**Comment by NiKiZe** [comment]
Test this:
Imgfetch in iPXE
Then chain into efi shell and run dir to list files.
You can also look at wimboot for listing files

**Comment by mcb30** [comment]
On 10/02/2022 23:29, tyne2022 wrote:
> I am using iPXE with an embedded system that uses U-Boot. I am able to 
> build iPXE as a EFI image for ARM64 and load it from U-Boot.
> 
> I would like to download an FPGA bitstream (generic binary file) over 
> HTTPS with iPXIE, get the address the image is held at, exit iPXE and 
> return to U-Boot, and finally configure the FPGA with the bitstream in 
> memory.
> 
> How can I get the address of an image (and size) of an image from 
> imgfetch? Is there a mechanism to pass this back to U-Boot?

There is no support for obtaining the address of an image (and nothing 
that guarantees that iPXE wouldn't choose to move the image at some 
point, though in practice this won't happen).  Also, relying on the 
image to still be present after iPXE has exited is not going to be 
reliable even if you were to create a mechanism for obtaining the image 
address.

A more viable approach is probably to have iPXE invoke the FPGA loading. 
  For example: if U-Boot were to expose its CLI interface via an EFI 
protocol then this makes it possible for an EFI application to invoke a 
U-boot command.  You could have iPXE invoke an EFI executable that loads 
the downloaded image from the EFI_SIMPLE_FILE_SYSTEM_PROTOCOL instance 
that iPXE exposes on the booted image's handle and then calls the U-Boot 
CLI command API to flash the FPGA.

If the U-Boot CLI interface were sufficiently generic, we could even 
have a "uboot" command added to iPXE, which could be used to invoke a 
U-Boot command directly from the iPXE shell (or script).

HTH,

Michael
